### ISSUE TITLE
fails to compile on macos with no such file or directory

### ISSUE THREAD

**Issue by emileakbarzadeh** [xtremekforever/swift-stm32c011-examples#1]
Hi,
Thanks for your work on this repo!

I'm having an issue with building the blink-mmio example on macos, getting a file not found error.

I am using the `org.swift.61202410271a` build of swift.

```
[1/1] Planning build
Building for production...
ar: @/Users/emile/Documents/swift-stm32c011-examples/blink-mmio/.build/armv6m-none-none-eabi/release/Application.product/Objects.LinkFileList: No such file or directory
[6/7] Archiving libApplication.a
make: *** [build] Error 1
```

`cat`ing the file shows it exists so I am very confused.

Thanks for your help. 

P.S. It would be super appreciated if you committed the files you used to generate the MMIO files with.

**Comment by xtremekforever** [comment]
Hmm interesting. Building has been working for me on macOS with the Swift 6.0.1 toolchain from Swift.org- so let me try the version you're using. I have been building this also with the main-snapshot 2024-10-25 on Ubuntu and it's been working too. So hopefully they didn't break something...

**Comment by xtremekforever** [comment]
Well, it's working fine for me. Not sure what's going on in your Mac:

```bash
blink-mmio$ swift --version
Apple Swift version 6.1-dev (LLVM f4b733c38006ec1, Swift f4bea5f7c0a7aa5)
Target: arm64-apple-macosx15.0
blink-mmio$ echo $TOOLCHAINS
org.swift.61202410271a
blink-mmio$ make
compiling...
swift build \
                --configuration release \
                --swift-sdks-path ../swift-sdks \
                --swift-sdk stm32c0xx \

[1/1] Planning build
Building for production...
[3/3] Write swift-version-3A674953373C380C.txt
Build complete! (0.55s)
linking...
arm-none-eabi-gcc ./.build/release/libApplication.a -o ./.build/release/blink.elf \
                -mcpu=cortex-m0plus -mthumb -mfloat-abi=soft \
                -ffreestanding -fdata-sections -ffunction-sections -fno-builtin -fno-common -fshort-enums -std=gnu11 \
                --specs=nano.specs --specs=nosys.specs \
                -Wl,--gc-sections \
                -T./Sources/Support/STM32C011F6.ld
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-closer.o): in function `_close_r':
closer.c:(.text._close_r+0xc): warning: _close is not implemented and will always fail
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-closer.o): note: the message above does not take linker garbage collection into account
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-signalr.o): in function `_getpid_r':
signalr.c:(.text._getpid_r+0x2): warning: _getpid is not implemented and will always fail
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-signalr.o): note: the message above does not take linker garbage collection into account
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-signalr.o): in function `_kill_r':
signalr.c:(.text._kill_r+0xe): warning: _kill is not implemented and will always fail
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-signalr.o): note: the message above does not take linker garbage collection into account
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-lseekr.o): in function `_lseek_r':
lseekr.c:(.text._lseek_r+0x10): warning: _lseek is not implemented and will always fail
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-lseekr.o): note: the message above does not take linker garbage collection into account
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-readr.o): in function `_read_r':
readr.c:(.text._read_r+0x10): warning: _read is not implemented and will always fail
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-readr.o): note: the message above does not take linker garbage collection into account
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-writer.o): in function `_write_r':
writer.c:(.text._write_r+0x10): warning: _write is not implemented and will always fail
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-writer.o): note: the message above does not take linker garbage collection into account
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-getentropyr.o): in function `_getentropy_r':
getentropyr.c:(.text._getentropy_r+0xe): warning: _getentropy is not implemented and will always fail
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/lib/thumb/v6-m/nofp/libc_nano.a(libc_a-getentropyr.o): note: the message above does not take linker garbage collection into account
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: warning: /Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/thumb/v6-m/nofp/crtn.o: missing .note.GNU-stack section implies executable stack
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
/Applications/ArmGNUToolchain/13.3.rel1/arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: warning: ./.build/release/blink.elf has a LOAD segment with RWX permissions
Extracting lst file...
arm-none-eabi-objdump -S .build/release/blink.elf > ./.build/release/blink.lst
Extracting map file...
arm-none-eabi-objdump -t .build/release/blink.elf | swift demangle > ./.build/release/blink.map
Extracting & printing sizes...
arm-none-eabi-size .build/release/blink.elf -A > ./.build/release/blink.size
   text    data     bss     dec     hex filename
    552     192       0     744     2e8 .build/release/blink.elf
Extracting bin file...
arm-none-eabi-objcopy -O binary .build/release/blink.elf ./.build/release/blink.bin
-rwxr-xr-x@ 1 xtremek  staff  744 Oct 30 12:36 ./.build/release/blink.bin
```

Arguably there are a lot of warnings from the linker, but I haven't seen that they break anything and I haven't had a chance to look into them. But, I can upload the SVD file I used with MMIO along with a basic README if that would help :)

**Comment by xtremekforever** [comment]
@emileakbarzadeh I added the SVD file for the STM32C011 and some instructions here: https://github.com/xtremekforever/swift-stm32c011-examples/tree/main/scripts

**Comment by emileakbarzadeh** [comment]
Thanks for writing those docs and providing the SVD, blink now works beautifully on my custom board (compiling with ubuntu).

After banging my head against a wall for a day, I've figured out the cause of the macos build issue.

The swift builder uses the `ar` utility to create packages, and uses a feature where the @ symbol indicates files whose contents are to be read as arguments. This function, however, is only included on the GNU version of `ar` NOT the version included in the BSD base of macOS, which includes an `ar` utility that doesn't use @ .

To resolve this issue, I needed to run 
``` bash
brew install binutils
echo 'export PATH="/opt/homebrew/opt/binutils/bin:$PATH"' >> /Users/<username>/.bash_profile
```

This installs the GNU version and puts it at the top of the path so it is used first.

I'm not sure if I haven't followed the instructions properly for installing swift or if this is an oversight from the swift developers.

Thank you @xtremekforever for your great work with this repo!
